src/app/components/settings/settings.rs

```diff
@@ -0,0 +1,271 @@
+                        1 => "160",
+                        2 => "320",
+                        _ => unreachable!(),
+                    }
+                    .to_variant()
+                })
+            })
+            .build();
+
+        let alsa_device = widget.alsa_device.downcast_ref::<gtk::Entry>().unwrap();
+        settings.bind("alsa-device", alsa_device, "text").build();
+
+        let audio_backend = widget
+            .audio_backend
+            .downcast_ref::<libadwaita::ComboRow>()
+            .unwrap();
+        settings
+            .bind("audio-backend", audio_backend, "selected")
+            .mapping(|variant, _| {
+                variant.str().map(|s| {
+                    match s {
+                        "pulseaudio" => 0,
+                        "alsa" => 1,
+                        _ => unreachable!(),
+                    }
+                    .to_value()
+                })
+            })
+            .set_mapping(|value, _| {
+                value.get::<u32>().ok().map(|u| {
+                    match u {
+                        0 => "pulseaudio",
+                        1 => "alsa",
+                        _ => unreachable!(),
+                    }
+                    .to_variant()
+                })
+            })
+            .build();
+
+        let ap_port = widget.ap_port.downcast_ref::<gtk::Entry>().unwrap();
+        settings
+            .bind("ap-port", ap_port, "text")
+            .mapping(|variant, _| variant.get::<u32>().map(|s| s.to_value()))
+            .set_mapping(|value, _| value.get::<u32>().ok().map(|u| u.to_variant()))
+            .build();
+
+        let theme = widget.theme.downcast_ref::<libadwaita::ComboRow>().unwrap();
+        settings
+            .bind("prefers-dark-theme", theme, "selected")
+            .mapping(|variant, _| {
+                variant
+                    .get::<bool>()
+                    .map(|prefer_dark| if prefer_dark { 1 } else { 0 }.to_value())
+            })
+            .set_mapping(|value, _| value.get::<u32>().ok().map(|u| (u == 1).to_variant()))
+            .build();
+    }
+
+    fn connect_theme_select(&self) {
+        let widget = imp::SettingsWindow::from_instance(self);
+        let theme = widget.theme.downcast_ref::<libadwaita::ComboRow>().unwrap();
+        theme.connect_selected_notify(|theme| {
+            let prefers_dark_theme = theme.selected() == 1;
+            let manager = libadwaita::StyleManager::default();
+
+            manager.set_color_scheme(if prefers_dark_theme {
+                libadwaita::ColorScheme::PreferDark
+            } else {
+                libadwaita::ColorScheme::PreferLight
+            });
+        });
+    }
+
+    fn connect_close<F>(&self, on_close: F)
+    where
+        F: Fn(SpotSettings) + 'static,
+    {
+        let pref_window = self.upcast_ref::<libadwaita::PreferencesWindow>();
+        let window = pref_window.upcast_ref::<libadwaita::Window>();
```

Comment:
think you can upcast_ref to a window in one go, no need to cast to a preferences window first